src/Shared/PackageManagers/ProjectManagerFactory.cs

```diff
@@ -8,55 +8,153 @@ namespace Microsoft.CST.OpenSource.PackageManagers
-        /// <param name="httpClientFactory"> The <see cref="IHttpClientFactory"/> for the project manager to use for making Http Clients to make web requests.</param>
-        /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
-        /// <returns> BaseProjectManager object </returns>
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        public delegate BaseProjectManager? ConstructProjectManager(string destinationDirectory = ".");
+
+        /// <summary>
+        /// The dictionary of project managers.
+        /// </summary>
+        private readonly Dictionary<string, ConstructProjectManager> _projectManagers;
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="overrideManagers">Use a custom set of managers to override the defaults with.</param>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        public ProjectManagerFactory(Dictionary<string, ConstructProjectManager> overrideManagers, IHttpClientFactory? httpClientFactory = null) : this(httpClientFactory)
         {
-            if (projectManagers.Count == 0)
-            {
-                projectManagers.AddRange(typeof(BaseProjectManager).Assembly.GetTypes()
-               .Where(type => type.IsSubclassOf(typeof(BaseProjectManager))));
-            }
-            // Use reflection to find the correct package management class
-            Type? downloaderClass = projectManagers
-               .Where(type => type.Name.Equals($"{purl.Type}ProjectManager",
-                                               StringComparison.InvariantCultureIgnoreCase))
-               .FirstOrDefault();
-
-            if (downloaderClass != null)
+            overrideManagers.ToList().ForEach(pair => SetManager(pair.Key, pair.Value));
+        }
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        public ProjectManagerFactory(IHttpClientFactory? httpClientFactory = null)
+        {
+            _projectManagers = CreateDefaultManagers(httpClientFactory);
+        }
+
+        public bool UnsetManager(string lookup)
+        {
+            return _projectManagers.Remove(lookup);
+        }
+
+        private void SetManager(string lookup, ConstructProjectManager generator)
+        {
+            _projectManagers[lookup] = generator;
+        }
+
+        public void ClearManagers()
+        {
+            _projectManagers.Clear();
+        }
```

Comment:
SetManager will be used when we are overriding a constructor, such as for unit testing NuGet we want to override the constructor with a mocked ManagerPackageActions (Not yet implemented)

Reply:
> SetManager will be used when we are overriding a constructor, such as for unit testing NuGet we want to override the constructor with a mocked ManagerPackageActions

Could that be accomplished by passing the overrides into one of the constructors of this class instead? It looks like there is already a constructor on this class that takes overrides?

Reply:
We can remove these methods if we change the constructor to take a full set instead of a subset to replace, and then in fact declare that it is backed by an immutable dictionary.

Since we have a "GetDefault" that is public the caller can get the default and then replace what they want to replace and then call the ctor.

Reply:
I removed UnsetManager and ClearManagers. I can't think of a scenario we would use them.